### PR TITLE
Adding skip checks flag to the upgrade script

### PR DESCRIPTION
**Pull request by amitaekbote**
## High Level Description

Fix error handling for the upgrade script. 
Remove the workaround which we had earlier for trapping error codes, now all the individual statements have the fixes
Add functionality to skip checks

Related changes in the test infra:
https://github.com/dcos/dcos-test-utils/pull/8

## Related Issues
https://jira.mesosphere.com/browse/DCOS-17844 

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

### CODE REVIEW

File: dcos_installer/upgrade.py [hunk 1]
```diff
@@ -34,19 +34,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-set -o errexit -o nounset -o pipefail
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: why?

File: dcos_installer/upgrade.py [hunk 2]
```diff
@@ -58,6 +49,12 @@
     fi
 fi
 
+SKIP_CHECKS=false
+
+if [ $1 = "skip-checks" ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: quote the `"$1"`

File: dcos_installer/upgrade.py [hunk 3]
```diff
@@ -71,22 +68,24 @@
...
-        exit 1
-    fi
-
-    clusteroutput=$(dcos-diagnostics check cluster)
-    if [ $? -ne 0 ]; then
-        echo "Cannot proceed with upgrade, cluster checks failed"
-        echo $clusteroutput
-        exit 1
-    fi
+if [ "$SKIP_CHECKS" = false ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: `if [[ "$SKIP_CHECKS" = "false" ]]; then`

File: dcos_installer/upgrade.py [hunk 4]
```diff
@@ -71,22 +68,24 @@
...
+   # Check if the node has node/cluster checks and run them
+   if [ -f /opt/mesosphere/etc/dcos-diagnostics-runner-config.json ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: double brackets

File: dcos_installer/upgrade.py [hunk 5]
```diff
@@ -71,22 +68,24 @@
...
+      # command exists
+      output="$(dcos-diagnostics check node-poststart 2>&1)"
+      if [ $? -ne 0 ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: `if [[ $? != 0 ]]; then`

Comment (karlkfi) [inline dcos_installer/upgrade.py]: nevermind. `!=` is the string comparison. `-ne` is fine.

File: dcos_installer/upgrade.py [hunk 6]
```diff
@@ -114,17 +113,20 @@
...
-    sleep 1
-    let T=T-1
-done
-RETCODE=$?
-if [ $RETCODE -ne 0 ]; then
-    echo "Node upgrade not successful, checks failed"
-    echo $OUT
+if [ "$SKIP_CHECKS" = false ]; then
+    T=300
+    until OUT="$(dcos-diagnostics check node-poststart && dcos-diagnostics check cluster 2>&1)" || [[ T -eq 0 ]]; do
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: `[[ $T -le 0 ]]`

File: dcos_installer/upgrade.py [hunk 7]
```diff
@@ -71,22 +68,24 @@
...
+      # command exists
+      output="$(dcos-diagnostics check node-poststart 2>&1)"
+      if [ $? -ne 0 ]; then
+          echo "Cannot proceed with upgrade, node checks failed"
+          echo "$output"
+          exit 1
+      fi
+
+      clusteroutput="$(dcos-diagnostics check cluster 2>&1)"
+      if [ $? -ne 0 ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: `if [[ $? -ne 0 ]]; then`

File: dcos_installer/upgrade.py [hunk 8]
```diff
@@ -114,17 +113,20 @@
...
-T=300
-until OUT=$(dcos-diagnostics check node-poststart && dcos-diagnostics check cluster) || [[ T -eq 0 ]]; do
-    sleep 1
-    let T=T-1
-done
-RETCODE=$?
-if [ $RETCODE -ne 0 ]; then
-    echo "Node upgrade not successful, checks failed"
-    echo $OUT
+if [ "$SKIP_CHECKS" = false ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: `if [[ "$SKIP_CHECKS" = "false" ]]; then`

File: dcos_installer/upgrade.py [hunk 9]
```diff
@@ -114,17 +113,20 @@
...
+    T=300
+    until OUT="$(dcos-diagnostics check node-poststart && dcos-diagnostics check cluster 2>&1)" || [[ T -eq 0 ]]; do
+      sleep 1
+      let T=T-1
+    done
+    RETCODE=$?
+    if [ $RETCODE -ne 0 ]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: double brackets

File: dcos_installer/upgrade.py [hunk 10]
```diff
@@ -114,17 +113,20 @@
...
+       echo "Node upgrade not successful, checks failed"
+       echo "$OUT"
+    fi
+    exit $RETCODE
 fi
-exit $RETCODE
+exit 0
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: unneeded. all scripts exit zero by default unless otherwise exited.

**Comment by amitaekbote** [comment]
@mesosphere-mergebot bump-ee

**Comment by amitaekbote** [comment]
@mesosphere-mergebot bump-ee

**Comment by amitaekbote** [comment]
@mesosphere-deploy label ready for review

**Comment by amitaekbote** [comment]
@mesosphere-mergebot label ready for review

**Comment by karlkfi** [comment]
@amitaekbote I think you misunderstood me. I was using the true = false thing to force an error after an echo to show you how the error handling worked.

**Comment by amitaekbote** [comment]
@karlkfi yeah ignore it for now, I was playing with it.  thanks for checking back at the pr. I will ping once its in a reviewable shape.

**Comment by amitaekbote** [comment]
@karlkfi ready for review now

**Comment by amitaekbote** [comment]
@mesosphere-mergebot bump-ee

**Comment by amitaekbote** [comment]
Test with skip checks 
https://teamcity.mesosphere.io/viewLog.html?buildId=748314&

Test without skip checks
https://teamcity.mesosphere.io/viewLog.html?buildId=748359&

**Comment by amitaekbote** [comment]
@mesosphere-mergebot bump-ee

File: dcos_installer/upgrade.py [hunk 11]
```diff
@@ -58,35 +49,44 @@
...
-        exit 1
-    fi
-
-    clusteroutput=$(dcos-diagnostics check cluster)
-    if [ $? -ne 0 ]; then
-        echo "Cannot proceed with upgrade, cluster checks failed"
-        echo $clusteroutput
-        exit 1
-    fi
+if [[ "$SKIP_CHECKS" = false ]]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: compare strings to strings
`if [[ "$SKIP_CHECKS" = "false" ]]; then`

File: dcos_installer/upgrade.py [hunk 12]
```diff
@@ -58,35 +49,44 @@
...
+   # Check if the node has node/cluster checks and run them
+   if [ -f /opt/mesosphere/etc/dcos-diagnostics-runner-config.json ]; then
+      # command exists
+      if ! output="$(dcos-diagnostics check node-poststart 2>&1)"; then
+          echo "Cannot proceed with upgrade, node checks failed"
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: do you want these echoes to go to stdout or etderr? Usually erro messages should go to stderr, unless you have a good reason not to.
```
echo >&2 "Cannot proceed with upgrade, node checks failed"
echo >&2 "$output"
```

File: dcos_installer/upgrade.py [hunk 13]
```diff
@@ -58,35 +49,44 @@
...
+          echo "$output"
+          exit 1
+      fi
+
+      if ! clusteroutput="$(dcos-diagnostics check cluster 2>&1)"; then
+          echo "Cannot proceed with upgrade, cluster checks failed"
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: stderr?

File: dcos_installer/upgrade.py [hunk 14]
```diff
@@ -114,17 +114,19 @@
...
-    echo $OUT
+if [[ "$SKIP_CHECKS" = false ]]; then
+    T=300
+    until OUT="$(dcos-diagnostics check node-poststart && dcos-diagnostics check cluster 2>&1)" || [[ $T -eq 0 ]]; do
+      sleep 1
+      let T=T-1
+    done
+    RETCODE=$?
+    if [[ $RETCODE -ne 0 ]]; then
+       echo "Node upgrade not successful, checks failed"
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: stderr?

File: dcos_installer/upgrade.py (hunk 12 above)
Comment (amitaekbote) [inline dcos_installer/upgrade.py]: In the case of dcos-diagnostics check the output always goes to stdout, its a json blob. It was intentionally designed that way

File: dcos_installer/upgrade.py (hunk 13 above)
Comment (amitaekbote) [inline dcos_installer/upgrade.py]: see comment above

File: dcos_installer/upgrade.py (hunk 12 above)
Comment (karlkfi) [inline dcos_installer/upgrade.py]: That's not what I'm asking. I'm asking if you want the failure output of this upgrade script to go to stderr.

Comment (amitaekbote) [inline dcos_installer/upgrade.py]: my bad, i misunderstood, let me change that

Comment (amitaekbote) [inline dcos_installer/upgrade.py]: Updated the pr, the actual output goes to stderr, message goes to stdout

**Comment by amitaekbote** [comment]
@mesosphere-mergebot bump-ee

File: dcos_installer/upgrade.py [hunk 15]
```diff
@@ -58,35 +49,44 @@
     fi
 fi
 
+SKIP_CHECKS=false
+
+if [[ $# -ne 0 ]]; then
```
Comment (karlkfi) [inline dcos_installer/upgrade.py]: you can blend these into a single if statement.
`if [[ $# -ne 0 ]] && [[ "$1" = "skip-checks" ]]; then`

**Comment by amitaekbote** [comment]
@mesosphere-mergebot label ship it

**Comment by amitaekbote** [comment]
Waiting on flaky tests

**Pull request merged**